#166286479 Enhance event to handle recurrent events

albert85 opened the pull request:
#### What Does This PR Do?
It enhances event model to handle recurrent events

#### Description Of Task To Be Completed
- Enhance event model to include recurrent start and end date
- Implement bulk create of recurrent events

#### Any Background Context You Want To Provide?
N/A

#### How can this be manually tested?
- Pull the branch
- Run migration by running `python server/manage.py makemigrations` and `python server/manage.py migrate` to update existing model with Recurrent model created
- check the diagram below to see the additional variables to be included to the event model

#### What are the relevant pivotal tracker stories?
[#166286479](https://www.pivotaltracker.com/story/show/166286479)

#### Screenshot(s)
<img width="593" alt="Screenshot 2019-05-31 at 7 53 59 PM" src="https://user-images.githubusercontent.com/28366272/58728036-e1c05480-83dd-11e9-8073-9e3a4e360309.png">

olusoladavid commented on server/api/models.py:
```diff
@@ -191,6 +191,10 @@ def get_count(self):
         return "Category : {}" .format(self.name)
 
+class Recurrence(BaseInfo):
+    frequency = models.CharField(max_length=50, default=0,)
```
I think our field types here could use better precision. Frequency for example is an integer field, start_date and end_date are DateTime fields. For example:
```python
frequency = models.IntegerField(min_value=0)
# we may not need the zero default since the recurrence foreign
# key is nullable
```
See this [link](https://www.webforefront.com/django/modeldatatypesandvalidation.html) for more info

olusoladavid commented on server/api/models.py:
```diff
@@ -200,6 +204,7 @@ class Event(BaseInfo):
     venue = models.TextField()
     start_date = models.CharField(max_length=50)
     end_date = models.CharField(max_length=50)
+    recurrence = models.ForeignKey(Recurrence, on_delete=models.CASCADE, null=True)
```
We may want to `explicitly` (default by Django is a `model_set` back-reference) define a backward relationship for the Foreign model here by explicitly setting the related_name and related_query_name arguments. Something like:
```python
recurrence = models.ForeignKey(
    Recurrence, 
    on_delete=models.CASCADE,
    related_name="events",
    related_query_name="event",
    null=True
)
```

olusoladavid commented on server/api/models.py:
```diff
@@ -191,6 +191,10 @@ def get_count(self):
     def __str__(self):
         return "Category : {}" .format(self.name)
 
+class Recurrence(BaseInfo):
```
Great job on this implementation @albert85. Looking at what we have now, I think `RecurringEvent` will be a better name for this model so that it is immediately clear to future maintainers of this codebase what this model does. I apologize that this was not caught in the spec earlier.

olusoladavid commented on server/graphql_schemas/event/schema.py:
```diff
@@ -68,11 +70,15 @@ class Input:
         timezone = graphene.String(required=False)
         slack_channel = graphene.String(required=False)
+        frequency = graphene.Int(default_value=0, required=False)
+        recurring=graphene.Boolean(required=False, default_value=False)
```
A few things here:
- Please add white space before and after `=`
- I don't think we need default values for `frequency` and `recurring` fields, since they are not required.

olusoladavid commented on server/api/serializers.py:
```diff
@@ -24,14 +24,23 @@ class Meta:
         depth = 1
 
 
+class RecurrenceSerializer(serializers.ModelSerializer):
```
While these serializer classes are no longer in active usage for most models (we switched from Django REST to GraphQL), we could write this RecurrenceSerializer class better. There's no `creator` field in the Recurrence model for example and there is no `date` or `time` field.

olusoladavid commented on server/graphql_schemas/event/schema.py:
```diff
@@ -81,12 +87,22 @@ def create_event(category, user_profile, **input):
-        new_event = Event.objects.create(
+        if input.get('recurring'):
+            recurrent_event = generate_date_time(input.get('start_date'), input.get('end_date'), input.get('frequency'), input, user_profile, category, recurrence_event)
+            print(recurrent_event)
```
I don't think we need this `print()` statement here

olusoladavid commented on server/graphql_schemas/event/schema.py:
```diff
@@ -81,12 +87,22 @@ def create_event(category, user_profile, **input):
+        if input.get('recurring'):
+            recurrent_event = generate_date_time(input.get('start_date'), input.get('end_date'), input.get('frequency'), input, user_profile, category, recurrence_event)
+            print(recurrent_event)
+            new_event = Event.objects.bulk_create(recurrent_event)
```
This `new_event` variable will contain a list of events created. This is the same variable that will be returned to the calendar event builder and slack event notifier _which expect `new_event` to be a single Event instance_. This is because we are violating the single responsibility principle on this `create_event` function. I will suggest we have a different function that handles creation of recurring events.

olusoladavid commented on server/graphql_schemas/event/schema.py:
```diff
@@ -98,8 +114,21 @@ def mutate_and_get_payload(cls, root, info, **input):
                 user=info.context.user
             )
+
+            if not input.get('recurring'):
```
I don't think we need this `if` block. This is because at the point of usage of the input, we are eventually deleting/ignoring the non-needed input keys.

olusoladavid commented on server/graphql_schemas/event/schema.py:
```diff
@@ -98,8 +114,21 @@ def mutate_and_get_payload(cls, root, info, **input):
+                input["recurrence_start_date"] = input["start_date"]
+                input["recurrence_end_date"] = input["end_date"]
+            
+            recurrence_event = CreateEvent.create_recurrent_event(**input)
```
This routine should ideally only run if `recurring` is true in the GraphQL input payload

olusoladavid commented on server/graphql_schemas/utils/helpers.py:
```diff
@@ -215,3 +215,62 @@ def add_event_to_calendar(andela_user, event):
     return calendar.events().insert(
             calendarId='primary', body=event_details).execute()
+
+def validate_recurrence_date(input):
```
I don't think we need this function. It appears to be a duplication of effort and code. With a little refactoring, `validate_event_dates()` already solves for this.

olusoladavid commented on server/graphql_schemas/utils/helpers.py:
```diff
@@ -215,3 +215,62 @@ def add_event_to_calendar(andela_user, event):
+    return totalTime
+
+
+def generate_date_time(start_date, end_date, freq, input, creator, social_event, recurrence):
```
I think this function is more appropriately named `generate_recurring_events`. Also the API surface is large. Most of the arguments are either already part of the Event input or the Recurrence instance. So we could have a simpler function like: `generate_recurring_events(event_dict, recurrence)`.

olusoladavid commented on server/graphql_schemas/utils/helpers.py:
```diff
@@ -215,3 +215,62 @@ def add_event_to_calendar(andela_user, event):
+
+
+def generate_date_time(start_date, end_date, freq, input, creator, social_event, recurrence):
+    generatedTimeObject = []
```
Our naming convention uses `snake_case` for variable names. Please modify as applicable.

olusoladavid commented on server/graphql_schemas/utils/helpers.py:
```diff
@@ -215,3 +215,62 @@ def add_event_to_calendar(andela_user, event):
+def generate_date_time(start_date, end_date, freq, input, creator, social_event, recurrence):
+    generatedTimeObject = []
+    generatedEvent = []
+    start_date = str(start_date).split('+')[0]
```
Is there any reason why we are throwing timezone information away here? Also important to note is that not all date strings will have a `+` sign. It may be `-` sign depending on timezone offset. To convert date strings to datetime objects, we could have:
```python
recurrence_start_date = parser.parse(start_date)
recurrence_end_date = parser.parse(end_date)
```

olusoladavid commented on server/graphql_schemas/utils/helpers.py:
```diff
@@ -215,3 +215,62 @@ def add_event_to_calendar(andela_user, event):
+            "recurrence": recurrence
+        }
+        generatedEvent.append(Event(**event_series))
+    return 
```
This function doesn't seem to be returning anything

olusoladavid reviewed (commented):
Great job on this implementation @albert85. I made some comments on how we can make it better. Please check them out. Cheers 🥂

albert85 committed:
feat(recurring events): Expand the events model to cater for recurring
events

- Create a RecurrenceEvents Model
- Adjust Events Mutation to cater for recurring events

[Delivers #166286479]
